### DISCUSSION TITLE
What's the equivalent of `git diff`?

### DISCUSSION THREAD

**Discussion by vincentfiestada** [go-git/go-git#685] [Q&A] [answered]
How can I compare the [_working tree_](https://github.com/go-git/go-git/blob/v5.5.2/worktree.go#L36) to a [Commit](https://github.com/go-git/go-git/blob/v5.5.2/plumbing/object/commit.go#L33) or [Tree](https://github.com/go-git/go-git/blob/v5.5.2/plumbing/object/tree.go#L34) object? This would be the equivalent of running `git diff` in the console.

**Comment by pjbgf** [comment]
Here's an extract from a test in `plumbing/object` which compares two trees:

```go
func (s *TreeSuite) TestTreeDiff(c *C) {
	f := fixtures.ByURL("https://github.com/src-d/go-git.git").One()
	storer := filesystem.NewStorage(f.DotGit(), cache.NewObjectLRUDefault())
	commit, err := GetCommit(storer, plumbing.NewHash("89f8bda31d29767a6d6ba8f9d0dfb941d598e843"))
	c.Assert(err, IsNil)

	tree, err := commit.Tree()
	c.Assert(err, IsNil)

	parentCommit, err := commit.Parent(0)
	c.Assert(err, IsNil)

	parentTree, err := parentCommit.Tree()
	c.Assert(err, IsNil)

	ch, err := parentTree.Diff(tree)
	c.Assert(err, IsNil)

	c.Assert(ch, HasLen, 3)
	c.Assert(ch[0].From.Name, Equals, "examples/object_storage/main.go")
	c.Assert(ch[0].To.Name, Equals, "examples/storage/main.go")

	ch, err = parentTree.DiffContext(context.Background(), tree)
	c.Assert(err, IsNil)
	c.Assert(ch, HasLen, 3)
}
```

https://github.com/go-git/go-git/blob/master/plumbing/object/tree_test.go#L244-L269

With the resulting `ch` you probably could use the upstream [DiffPrettyText](https://github.com/sergi/go-diff/blob/facec63e78161d6d31a9c552a679e2287e925949/diffmatchpatch/diff.go#L1147) to transform that into text:

**Reply by aslakhellesoy** [reply]
@vincentfiestada did you find out how to do this?

I don't see how @pjbgf's example above builds a `Tree` from a `Worktree` - the code is doing it from a commit (`89f8bda31d29767a6d6ba8f9d0dfb941d598e843`)

**Comment by AriehSchneier** [comment, marked as answer]
Do you have the repo rather than the worktree? If so you could try (untested code written here):
```
	to = object.NewTreeRootNode(t)

	idx, err := repo.Storer.Index()
	if err != nil {
		return nil, err
	}
	// mindex is utils/merkletrie/index
	from := mindex.NewRootNode(idx)

	changes, err := merkletrie.DiffTree(from, to, diffTreeIsEquals)
```

**Reply by aslakhellesoy** [reply]
Thanks @AriehSchneier - that seems to work. If I understand correctly, the `from` object here represents the staging area, right? Similar to `git diff --cached`?

Do you know if it would be possible to make `from` represent the working directory instead?

I'm trying to get the diff between a commit (HEAD) and the working directory (where some files might not be added to staging yet).

**Reply by AriehSchneier** [reply]
Ah yeh, thats correct. I think what you are after is this function: https://github.com/go-git/go-git/blob/dd4e2b7f4b01e2aaafcf182d2884a186a6f37d21/worktree_status.go#L50

Depending on what you are after, you could either propose a change that exposes that function (probably just needs a clearer name and then can call directly into it), or you could try copy the code locally.

PS my code snippet above came from `diffTreeWithStaging`

**Reply by AriehSchneier** [reply]
Oh, I missed that the commit you are trying to compare is HEAD! `w.Status()` should do that: https://github.com/go-git/go-git/blob/dd4e2b7f4b01e2aaafcf182d2884a186a6f37d21/worktree_status.go#L35

**Reply by aslakhellesoy** [reply]
Yeah, I had a hunch I might have to use `w.Status()`. Unfortunately it's too slow for me (ref https://github.com/go-git/go-git/issues/181) - I have the black hole that is `node_modules` 😱 

Looks like I'll have to shell out to `git status --porcelain --untracked-files=no` instead.

**Reply by AriehSchneier** [reply]
I had a super quick look, it does the entire diff without taking into account untracked files, only afterwards does it then filter them out, hence why you are finding it slow with lots of untracked files...

**Comment by bluebrown** [comment]
I am looking for to ignore whitespace changes. With git CLI, this seems to work

```bash
git diff --ignore-blank-lines --ignore-all-space --ignore-cr-at-eol --stat
```

Is there a way to do it with this go module?
